chain-core/src/state/validator/nodejoin.rs

```diff
@@ -0,0 +1,83 @@
+use crate::state::account::{CouncilNode, Nonce, StakedStateAddress, StakedStateOpAttributes};
+use crate::tx::TransactionId;
+use parity_scale_codec::{Decode, Encode, Error, Input, Output};
+#[cfg(not(feature = "mesalock_sgx"))]
+use serde::{Deserialize, Serialize};
+#[cfg(not(feature = "mesalock_sgx"))]
+use std::fmt;
+
+/// Submits a proposal to add a council node:
+/// if there are less than max cap validators or the associated stake
+/// is more than the smallest one in the validator set,
+/// the validator set will be updated.
+///
+/// tx-validation should check that:
+/// - the address and the consensus_pubkey are not in the
+/// - the associated staked state is ok (not jailed etc.)
+/// - the bonded amount in the stake state is more than the minimal required one
+/// - the witness is correct
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[cfg_attr(not(feature = "mesalock_sgx"), derive(Serialize, Deserialize))]
+pub struct NodeJoinRequestTx {
+    pub nonce: Nonce,
+    pub address: StakedStateAddress,
+    pub attributes: StakedStateOpAttributes,
+    pub node_meta: CouncilNode,
+}
+
+impl Decode for NodeJoinRequestTx {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+        let nonce = Nonce::decode(input)?;
+        let address = StakedStateAddress::decode(input)?;
+        let attributes = StakedStateOpAttributes::decode(input)?;
+        let node_meta = CouncilNode::decode(input)?;
+
+        Ok(NodeJoinRequestTx {
+            nonce,
+            address,
+            attributes,
+            node_meta,
+        })
+    }
+}
+
+// TODO: size hint as node_meta needs more info
+impl Encode for NodeJoinRequestTx {
+    fn encode_to<EncOut: Output>(&self, dest: &mut EncOut) {
+        dest.push(&self.nonce);
+        dest.push(&self.address);
+        dest.push(&self.attributes);
+        dest.push(&self.node_meta);
```

Comment:
I feel that we might want to be able to extend `CouncilNode` in the future?

Reply:
The part around `CouncilNode` is unfinished e.g. with regards to TDBE. but yeah, it may be better to tag it nonetheless